working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,470 @@
+    But I do think it could be useful to offer this as a lint with a metadata
+    annotation for users that are more cautious, like:
+
+    ```dart
+    interface class NoExtend {}
+    @reopen class MySubclass extends NoExtend {}
+    ```
+
+    This proposal takes that option.
+
+Finally, to the actual proposal...
+
+## Syntax
+
+This proposal builds on the existing sealed types proposal so the grammar
+includes those changes. The full set of modifiers that can appear before a class
+or mixin are `abstract`, `sealed`, `base`, `interface`, and `final`. Some
+combinations don't make sense:
+
+*   `sealed` implies `abstract`, so they can't be combined.
+*   `sealed` implies non-extensibility, so can't be combined with `interface`
+    or `final`.
+*   `sealed` implies non-implementability, so can't be combined with `base` or
+    `final`.
+*   `base`, `interface`, and `final` all control the same two capabilities so
+    are mutually exclusive.
+*   `final` and `interface` imply that the type can't be used as a superclass,
+    but the reason for defining a mixin is so that it can be mixed in and become
+    a superclass, so we don't allow those on mixins.
+
+The remaining valid combinations are:
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
+
+mixin
+sealed mixin
+base mixin
+interface mixin
+final mixin
+```
+
+The grammar is:
+
+```
+classDeclaration ::=
+  classModifiers 'class' identifier typeParameters?
+  superclass? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+  | classModifiers 'class' mixinApplicationClass
+
+mixinDeclaration ::= mixinModifiers 'mixin' identifier typeParameters?
+  ('on' typeNotVoidList)? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+
+classModifiers ::= 'sealed' | 'abstract'? typeModifiers?
+mixinModifiers ::= 'sealed' | typeModifiers?
+typeModifiers  ::= 'base' | 'interface' | 'final'
+```
+
+### Static semantics
+
+It is a compile-time error to:
+
+*   Extend a class marked `interface` or `final` outside of the library where it
+    is defined.
+
+*   Implement a type marked `base` or `final` outside of the library where it is
+    defined.
+
+*   Extend or mix in a type marked `base` outside of the library where it is
+    defined without also being marked `base` or `final`. *This ensures that a
+    subtype can't escape the `base` restriction of its supertype by offering its
+    _own_ interface that could then be implemented without inheriting the
+    concrete implementation from the supertype.*
+
+*   Mix in a class marked `sealed`, `base`, `interface`, or `final`. *We want to
+    eventually move away from classes as mixins. We don't want to break existing
+    uses of classes as mixins but since no existing code is using these
+    modifiers, we can prevent classes using those modifiers from also being used
+    as mixins.*
```

Comment:
Ahhh, there it is.

If we don't think we can get there, I'd be fine with just marking `class`es that can be mixed in as `mixin class`.

Reply:
> Maybe we should embrace that some types are both mixins and classes, and allow you to declare them as mixin class.

We have talked about that before in earlier discussions around all of the various combinations of capabilities, but I don't think we ever found any real use cases for it. Given how confusing mixins already are, I think it makes the most sense to just keep them separate.

You can always define a mixin and then a mixin application class of it if you want to give users both affordances.